### PR TITLE
Add a partial paginator to prevent SQL COUNT queries

### PR DESCRIPTION
**Pull request by meyerbaptiste**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

Recently we had to profile an API Platform application. This one had quite long response times because of the Doctrine paginator and more specifically the SQL `COUNT()` query it executes. Based on theses observations, I created a partial paginator which doesn't execute this query and therefore returns only the previous and the next page.

Next step: next and previous page with a cursor (Facebook, Twitter, etc. like).

Cc @thomasglachant

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/AbstractPaginator.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\Orm;
+
+use ApiPlatform\Core\DataProvider\PartialPaginatorInterface;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+
+abstract class AbstractPaginator implements \IteratorAggregate, PartialPaginatorInterface
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: FYI,

`abstract class AbstractPaginator implements PartialPaginatorInterface, \IteratorAggregate` => fatal error
`abstract class AbstractPaginator implements \IteratorAggregate, PartialPaginatorInterface` => it works

🤔 🤣

File: src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php [hunk 2]
```diff
@@ -33,8 +33,10 @@ public function supportsResult(string $resourceClass, string $operationName = nu
 
     /**
      * @param QueryBuilder $queryBuilder
+     * @param string       $resourceClass
+     * @param string|null  $operationName
      *
      * @return mixed
      */
-    public function getResult(QueryBuilder $queryBuilder);
+    public function getResult(QueryBuilder $queryBuilder/*, string $resourceClass, string $operationName = null*/);
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php]: Why?

**Comment by soyuka** [comment]
Shouldn't you change the `CollectionDataProvider` and the `SubresourceDataProvider` also?

**Comment by bendavies** [comment]
I also noticed this performance hit.  
One thing that could be improved is removing left joins on the count query, which dramatically improved performance for my use case.

File: src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php (hunk 2 above)
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php]: To keep consistency with `QueryResultCollectionExtensionInterface` but yes, maybe it's useless...

**Comment by soyuka** [comment]
:+1: @bendavies thought about this too but the doctrine paginator is a dark place to change things like this (it works on the same basis QueryBuilder for all the queries).

**Comment by bendavies** [comment]
@soyuka i can't see much apart from writing our own Paginator or decorating it and overriding `count`

**Comment by soyuka** [comment]
> @soyuka i can't see much apart from writing our own Paginator or decorating it and overriding count

I meant to remove `leftJoins` from the `COUNT` query but not from the end query. Anyway this pull request gives the ability to skip the `COUNT` :).

File: src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php (hunk 2 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php]: Keep it, at least we have them if we need it later :P.

**Comment by meyerbaptiste** [comment]
If I remember correctly, on the project we profiled there was already a decorator to remove in some cases the `LEFT JOIN` part of the query but that did not prevent a long execution time.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -96,13 +96,16 @@ public function getConfigTreeBuilder()
                             ->addDefaultsIfNotSet()
                             ->children()
                                 ->booleanNode('enabled')->defaultTrue()->info('To enable or disable pagination for all resource collections by default.')->end()
+                                ->booleanNode('partial')->defaultFalse()->info('')->end()
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Can you add an `info`?

Comment (meyerbaptiste) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Oops!

File: src/Hal/Serializer/CollectionNormalizer.php [hunk 4]
```diff
@@ -68,15 +69,17 @@ public function normalize($object, $format = null, array $context = [])
         $resourceClass = $this->resourceClassResolver->getResourceClass($object, $context['resource_class'] ?? null, true);
         $context = $this->initContext($resourceClass, $context);
         $parsed = IriHelper::parseIri($context['request_uri'] ?? '/', $this->pageParameterName);
-        $paginated = $isPaginator = $object instanceof PaginatorInterface;
 
-        $currentPage = $lastPage = $itemsPerPage = null;
-        if ($isPaginator) {
+        $currentPage = $lastPage = $itemsPerPage = $pageTotalItems = null;
```
Comment (soyuka) [inline src/Hal/Serializer/CollectionNormalizer.php]: Whew, lots of variables initiated to `nil`, is this required?

File: src/Hydra/Serializer/PartialCollectionViewNormalizer.php [hunk 5]
```diff
@@ -50,15 +51,16 @@ public function normalize($object, $format = null, array $context = [])
...
 
-        $currentPage = $lastPage = null;
-        if ($paginated = $object instanceof PaginatorInterface) {
-            $currentPage = $object->getCurrentPage();
-            $lastPage = $object->getLastPage();
-
-            if (1. === $currentPage && 1. === $lastPage) {
-                // Consider the collection not paginated if there is only one page
-                $paginated = false;
+        $currentPage = $lastPage = $itemsPerPage = $pageTotalItems = null;
```
Comment (soyuka) [inline src/Hydra/Serializer/PartialCollectionViewNormalizer.php]: same I wonder why is this needed

File: src/Bridge/Doctrine/Orm/AbstractPaginator.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function __construct(DoctrinePaginator $paginator)
+    {
+        $query = $paginator->getQuery();
+
+        if (null === ($firstResult = $query->getFirstResult()) || null === $maxResults = $query->getMaxResults()) {
+            throw new InvalidArgumentException(sprintf('"%1$s::setFirstResult()" or/and "%1$s::setMaxResults()" was/were not applied to the query.', Query::class));
+        }
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: I saw the test below but I don't get why we need such message (as we didn't needed it before)?

Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: Because these two methods can potentially return `null`. See the Doctrine PHPDoc:

> Returns NULL if {link setFirstResult} was not applied to this query.
> Returns NULL if {link setMaxResults} was not applied to this query.

We always call `Query::setFirstResult()` and `Query::setMaxResults()` before instantiating our `Paginator` class, it's why we never had the bug.

File: src/Hal/Serializer/CollectionNormalizer.php (hunk 4 above)
Comment (meyerbaptiste) [inline src/Hal/Serializer/CollectionNormalizer.php]: Because of this?

```shell
------ --------------------------------------------- 
  Line   src/Hal/Serializer/CollectionNormalizer.php  
 ------ --------------------------------------------- 
  87     Undefined variable: $currentPage             
  92     Undefined variable: $lastPage                
  97     Undefined variable: $currentPage             
  98     Undefined variable: $currentPage             
  101    Undefined variable: $currentPage             
  101    Undefined variable: $pageTotalItems          
  101    Undefined variable: $itemsPerPage            
  102    Undefined variable: $currentPage             
  123    Undefined variable: $itemsPerPage            
 ------ --------------------------------------------- 
 ------ ---------------------------------------------------------- 
  Line   src/Hydra/Serializer/PartialCollectionViewNormalizer.php  
 ------ ---------------------------------------------------------- 
  75     Undefined variable: $currentPage                          
  80     Undefined variable: $lastPage                             
  85     Undefined variable: $currentPage                          
  86     Undefined variable: $currentPage                          
  89     Undefined variable: $currentPage                          
  89     Undefined variable: $pageTotalItems                       
  89     Undefined variable: $itemsPerPage                         
  90     Undefined variable: $currentPage                          
 ------ ---------------------------------------------------------- 
```

File: src/Bridge/Doctrine/Orm/AbstractPaginator.php (hunk 6 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: > We always call Query::setFirstResult() and Query::setMaxResults() before instantiating our Paginator class, it's why we never had the bug.

Yes that's the reason why we don't need the test no? I mean we're still always using `setFirstResult` and `setMaxResults` or I miss something.

Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: Yes but this is a library and this class is not internal so users are free to use it as they want. IMO we have to prevent a fatal error by preventing this specific use case.

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: Okay wasn't sure this was the reason but I thought about it! Thanks for clarifying!

**Comment by Simperfit** [comment]
@meyerbaptiste could you please rebase ? this should be ok to merge, @dunglas could you please look at this ?

**Comment by Simperfit** [comment]
@meyerbaptiste could you rebase again, so we can merge it ?

File: src/Bridge/Doctrine/Orm/AbstractPaginator.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public function getItemsPerPage(): float
+    {
+        return (float) $this->maxResults;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator()
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: You can add the `\Traversable` return type even if not in the parent class.

File: phpstan.neon [hunk 8]
```diff
@@ -14,3 +14,8 @@ parameters:
         - '#Parameter \#2 \$dqlPart of method Doctrine\\ORM\\QueryBuilder::add\(\) expects Doctrine\\ORM\\Query\\Expr\\Base, Doctrine\\ORM\\Query\\Expr\\Join\[\] given#' # Fixed in Doctrine's master
         - '#Call to an undefined method Doctrine\\Common\\Persistence\\ObjectManager::getConnection\(\)#'
         - '#Parameter \#1 \$callable of static method Doctrine\\Common\\Annotations\\AnnotationRegistry::registerLoader\(\) expects callable, mixed\[\] given#'
+        - '#Method ApiPlatform\\Core\\Bridge\\Doctrine\\Orm\\Extension\\QueryResult(Item|Collection)ExtensionInterface::getResult\(\) invoked with 3 parameters, 1 required#'
+        - '#Undefined variable: \$currentPage#'
```
Comment (dunglas) [inline phpstan.neon]: These `Undefined variable ` look like code smell.

File: src/Bridge/Doctrine/Orm/AbstractPaginator.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        return $this->iterator ?? $this->iterator = $this->paginator->getIterator();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function count()
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/AbstractPaginator.php]: Same here (`: int`)

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 10]
```diff
@@ -108,14 +115,55 @@ public function supportsResult(string $resourceClass, string $operationName = nu
...
      */
-    public function getResult(QueryBuilder $queryBuilder)
+    public function getResult(QueryBuilder $queryBuilder/*, string $resourceClass, string $operationName = null*/)
     {
+        $resourceClass = $operationName = null;
+
+        if (func_num_args() >= 2) {
+            $resourceClass = func_get_arg(1);
+        } else {
+            @trigger_error(sprintf('Method %s() will have a 2nd `string $resourceClass` argument in version 3.0. Not defining it is deprecated since 2.1.', __METHOD__), E_USER_DEPRECATED);
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: 2.2 (same everywhere)

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 11]
```diff
@@ -108,14 +115,55 @@ public function supportsResult(string $resourceClass, string $operationName = nu
...
+        if (func_num_args() >= 3) {
+            $operationName = func_get_arg(2);
+        } else {
+            @trigger_error(sprintf('Method %s() will have a 3rd `string $operationName = null` argument in version 3.0. Not defining it is deprecated since 2.1.', __METHOD__), E_USER_DEPRECATED);
+        }
+
         $doctrineOrmPaginator = new DoctrineOrmPaginator($queryBuilder, $this->useFetchJoinCollection($queryBuilder));
         $doctrineOrmPaginator->setUseOutputWalkers($this->useOutputWalkers($queryBuilder));
 
+        $resourceMetadata = null !== $resourceClass ? $this->resourceMetadataFactory->create($resourceClass) : null;
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: `null === $resourceClass ? null : $this->resourceMetadataFactory->create($resourceClass)` (negation not needed).

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php [hunk 12]
```diff
@@ -108,14 +115,55 @@ public function supportsResult(string $resourceClass, string $operationName = nu
...
+
         return new Paginator($doctrineOrmPaginator);
     }
 
+    private function isPartialPaginationEnabled(Request $request = null, ResourceMetadata $resourceMetadata = null, string $operationName = null): bool
+    {
+        $enabled = $this->partial;
+        $clientEnabled = $this->clientPartial;
+
+        if ($resourceMetadata) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: `null !==` (same for all tests below)

File: src/Bridge/Doctrine/Orm/Paginator.php [hunk 13]
```diff
@@ -14,95 +14,32 @@
...
 use ApiPlatform\Core\DataProvider\PaginatorInterface;
-use Doctrine\ORM\Tools\Pagination\Paginator as DoctrineOrmPaginator;
 
 /**
  * Decorates the Doctrine ORM paginator.
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  */
-final class Paginator implements \IteratorAggregate, PaginatorInterface
+final class Paginator extends AbstractPaginator implements PaginatorInterface
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Paginator.php]: `implements PaginatorInterface` can be removed (already implemented by the `abstract` class).

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 14]
```diff
@@ -116,13 +116,16 @@ public function getConfigTreeBuilder()
                             ->addDefaultsIfNotSet()
                             ->children()
                                 ->booleanNode('enabled')->defaultTrue()->info('To enable or disable pagination for all resource collections by default.')->end()
+                                ->booleanNode('partial')->defaultFalse()->info('To enable or disable partial pagination for all resource collections by default when pagination is enabled.')->end()
                                 ->booleanNode('client_enabled')->defaultFalse()->info('To allow the client to enable or disable the pagination.')->end()
                                 ->booleanNode('client_items_per_page')->defaultFalse()->info('To allow the client to set the number of items per page.')->end()
+                                ->booleanNode('client_partial')->defaultFalse()->info('')->end()
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: `info('')` must be filled (BTW, I'm not sure that it's really necessary to be able to set this option client-side, but as it's implemented, keep it)

File: src/Hal/Serializer/CollectionNormalizer.php [hunk 15]
```diff
@@ -68,15 +69,17 @@ public function normalize($object, $format = null, array $context = [])
         $resourceClass = $this->resourceClassResolver->getResourceClass($object, $context['resource_class'] ?? null, true);
         $context = $this->initContext($resourceClass, $context);
         $parsed = IriHelper::parseIri($context['request_uri'] ?? '/', $this->pageParameterName);
-        $paginated = $isPaginator = $object instanceof PaginatorInterface;
 
-        $currentPage = $lastPage = $itemsPerPage = null;
```
Comment (dunglas) [inline src/Hal/Serializer/CollectionNormalizer.php]: I think those changes introduce the PHPStan errors. It should probably be reverted.

File: src/Bridge/Doctrine/Orm/Paginator.php (hunk 13 above)
Comment (meyerbaptiste) [reply src/Bridge/Doctrine/Orm/Paginator.php]: Nop, `AbstractPaginator` implements `PartialPaginatorInterface`.

File: src/Hal/Serializer/CollectionNormalizer.php (hunk 15 above)
Comment (meyerbaptiste) [reply src/Hal/Serializer/CollectionNormalizer.php]: Yep, see https://github.com/api-platform/core/pull/1292#discussion_r130375281

Comment (soyuka) [reply src/Hal/Serializer/CollectionNormalizer.php]: Feels weird to me to declare empty variables but maybe it's better... Your call @dunglas.

File: src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php (hunk 12 above)
Comment (meyerbaptiste) [reply src/Bridge/Doctrine/Orm/Extension/PaginationExtension.php]: Should I do it after yesterday's conversation? 😛

File: src/Bridge/Doctrine/Orm/Paginator.php (hunk 13 above)
Comment (dunglas) [reply src/Bridge/Doctrine/Orm/Paginator.php]: Not keep it as is :) Sorry.

File: src/Hal/Serializer/CollectionNormalizer.php (hunk 15 above)
Comment (dunglas) [reply src/Hal/Serializer/CollectionNormalizer.php]: It makes the execution path more predictible for static analysis tools (and even for humans IMO).

Comment (soyuka) [reply src/Hal/Serializer/CollectionNormalizer.php]: Humans > Robots

**Comment by meyerbaptiste** [comment]
Done @dunglas.

File: src/Hal/Serializer/CollectionNormalizer.php [hunk 16]
```diff
@@ -68,15 +69,18 @@ public function normalize($object, $format = null, array $context = [])
...
-        if ($isPaginator) {
+        $currentPage = $lastPage = $itemsPerPage = $pageTotalItems = null;
+        if ($paginated = $isPaginator = $object instanceof PartialPaginatorInterface) {
+            if ($object instanceof PaginatorInterface) {
+                $paginated = 1. !== $lastPage = $object->getLastPage();
+            } else {
+                $lastPage = null;
```
Comment (soyuka) [inline src/Hal/Serializer/CollectionNormalizer.php]: Is this one needed if we declared it above?

File: src/Hydra/Serializer/PartialCollectionViewNormalizer.php [hunk 17]
```diff
@@ -50,15 +51,17 @@ public function normalize($object, $format = null, array $context = [])
...
+        if ($paginated = $object instanceof PartialPaginatorInterface) {
+            if ($object instanceof PaginatorInterface) {
+                $paginated = 1. !== $lastPage = $object->getLastPage();
+            } else {
+                $lastPage = null;
```
Comment (soyuka) [inline src/Hydra/Serializer/PartialCollectionViewNormalizer.php]: same now that it's declared above

File: src/Hal/Serializer/CollectionNormalizer.php (hunk 16 above)
Comment (meyerbaptiste) [reply src/Hal/Serializer/CollectionNormalizer.php]: Done.

File: src/Hydra/Serializer/PartialCollectionViewNormalizer.php (hunk 17 above)
Comment (meyerbaptiste) [reply src/Hydra/Serializer/PartialCollectionViewNormalizer.php]: Done.

**Comment by soyuka** [comment]
Thanks @meyerbaptiste  :D

**Comment by thomasglachant** [comment]
Thanks a lot, great job 👍
